### PR TITLE
[browser] Integrate DevServer into WasmAppHost

### PR DESCRIPTION
**Pull request by maraf**
- Integrate (copy) Blazor DevServer into WasmAppHost
  - DevServer origins
    - Copied from Blazor DevServer without any change
      - `ComponentsWebAssemblyApplicationBuilderExtensions.cs`
      - `ContentEncodingNegotiator.cs`
      - `WebAssemblyNetDebugProxyAppBuilderExtensions.cs`
    - `DevServer.cs` is mix between `WebServer` and [asptnetcore DevServer/Program.cs](https://github.com/dotnet/aspnetcore/blob/ba87dacb68a221ba9e04036b073b40c4aabc3f43/src/Components/WebAssembly/DevServer/src/Server/Program.cs)
    - `DevServerStartup.cs` mix between `WebServerStartup` and [asptnetcore DevServer/Startup.cs](https://github.com/dotnet/aspnetcore/blob/ba87dacb68a221ba9e04036b073b40c4aabc3f43/src/Components/WebAssembly/DevServer/src/Server/Startup.cs)
    - `DevServerOptions.cs` is inspired by `WebServerOptions`
    - `DebugProxyLauncher.cs` changed folder where it expects debug proxy
  - The WasmAppHost still remains in workload, because of the console template
  - The WasmAppHost is also included in the Wasm SDK pack for running without workload from Wasm SDK
  - If we finish the console template in .NET 8, we can remove WasmAppHost from the workload
- The BrowserHost switches between WebServer and DevServer based on static web assets manifest presence
  - We probably could remove WebServer later
  - The solution still uses runtime config
- Make `HTMLPath` optional
- Set run parameters for Wasm SDK
  - It's in the `Browser.targets` because that is the topmost place where `OutputPath` is correctly set. Unfortunatelly it's after the default run parameters are set. So it must override it
  - It uses `_WebAssemblyUserRunParameters` to preserve user defined values. The property will be set in SDK https://github.com/dotnet/sdk/commit/569f220250fc11c60116bf5b3e5e57a3180e0e94
  - It checks for "blazor-devserver.dll" to preserve Blazor setting. We could use different check once the `UsingMicrosoftNETSdkBlazorWebAssembly` is removed from Wasm SDK
- ~Until resolved, TestAppScenarios pass `GenerateRuntimeConfigurationFiles=true` as command line argument to override [Microsoft.NET.Sdk.WebAssembly.Browser.targets#L64
](https://github.com/dotnet/runtime/blob/51af14c745e1074340a605faad3ba1bf8a969868/src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets#L64)~
- Remove `GenerateRuntimeConfigurationFiles=false` from Wasm SDK, since the WasmAppHost needs it and Blazor SDK https://github.com/dotnet/sdk/pull/34177 change already flown to installer 
- DebugProxy tested manually
- Tested that is doesn't interfere with Blazor's DevServer from nuget package

Contributes to https://github.com/dotnet/runtime/issues/70762

### CODE REVIEW

File: src/mono/wasm/testassets/WasmBasicTestApp/runtimeconfig.template.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (maraf) [inline src/mono/wasm/testassets/WasmBasicTestApp/runtimeconfig.template.json:1]: @radical Is there an explicity step needed to for this template to be transformed into runtimeconfig? I couldn't found anything, but when this app is built, the template is copied as is to the output

Comment (maraf) [reply src/mono/wasm/testassets/WasmBasicTestApp/runtimeconfig.template.json:1]: Nevermind https://github.com/dotnet/runtime/blob/51af14c745e1074340a605faad3ba1bf8a969868/src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets#L64

**Comment by maraf** [comment]
Debugger tests are not actually using code changes from this PR

File: src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        options.FileProvider = webRootFileProvider;
+        var contentTypeProvider = new FileExtensionContentTypeProvider();
+        AddMapping(contentTypeProvider, ".dll", MediaTypeNames.Application.Octet);
+        AddMapping(contentTypeProvider, ".webcil", MediaTypeNames.Application.Octet);
+        // We unconditionally map pdbs as there will be no pdbs in the output folder for
+        // release builds unless BlazorEnableDebugging is explicitly set to true.
+        AddMapping(contentTypeProvider, ".pdb", MediaTypeNames.Application.Octet);
+        AddMapping(contentTypeProvider, ".br", MediaTypeNames.Application.Octet);
+        AddMapping(contentTypeProvider, ".dat", MediaTypeNames.Application.Octet);
+        AddMapping(contentTypeProvider, ".blat", MediaTypeNames.Application.Octet);
```
Comment (pavelsavara) [inline src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs:97]: We could remove `.blat`

File: src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+        var webHostEnvironment = builder.ApplicationServices.GetRequiredService<IWebHostEnvironment>();
+
+        var options = CreateStaticFilesOptions(webHostEnvironment.WebRootFileProvider);
+
+        builder.MapWhen(
+            ctx => ctx.Request.Path.StartsWithSegments(pathPrefix, out var rest)
+                && rest.StartsWithSegments("/_framework")
+                && !rest.StartsWithSegments("/_framework/blazor.server.js")
+                && !rest.StartsWithSegments("/_framework/blazor.web.js"),
```
Comment (pavelsavara) [inline src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs:43]: What we need here to make non-blazor also hot-reload enabled ?

File: src/mono/wasm/host/DevServer/DebugProxyLauncher.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    private static string LocateDebugProxyExecutable(IWebHostEnvironment environment)
+    {
+        if (string.IsNullOrEmpty(environment.ApplicationName))
+        {
+            throw new InvalidOperationException("IWebHostEnvironment.ApplicationName is required to be set in order to start the debug proxy.");
+        }
+        var assembly = Assembly.Load(environment.ApplicationName);
+        var debugProxyPath = Path.Combine(
+            Path.GetDirectoryName(assembly.Location)!,
+            "BlazorDebugProxy",
```
Comment (pavelsavara) [inline src/mono/wasm/host/DevServer/DebugProxyLauncher.cs:110]: `BlazorDebugProxy` do we have such thing ?

File: src/mono/wasm/host/DevServer/DevServerStartup.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+        app.UseWebAssemblyDebugging();
+
+        bool applyCopHeaders = configuration.GetValue<bool>("ApplyCopHeaders");
+
+        if (applyCopHeaders)
+        {
+            app.Use(async (ctx, next) =>
+            {
+                if (ctx.Request.Path.StartsWithSegments("/_framework") && !ctx.Request.Path.StartsWithSegments("/_framework/blazor.server.js") && !ctx.Request.Path.StartsWithSegments("/_framework/blazor.web.js"))
```
Comment (pavelsavara) [inline src/mono/wasm/host/DevServer/DevServerStartup.cs:43]: Do we need more files for non-blazor ?

File: src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs (hunk 2 above)
Comment (maraf) [reply src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs:97]: Theoretically. My longer term goal is to use WasmAppHost even for Blazor projects and than we would nee `.blat` from hosting .NET 7 Blazor.  I'm not sure yet if that is doable path. Leaving the extension there probably doesn't do any harm.

File: src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs (hunk 3 above)
Comment (maraf) [reply src/mono/wasm/host/DevServer/ComponentsWebAssemblyApplicationBuilderExtensions.cs:43]: There is more work on that one and I don't know all the details yet. There is a module in SDK https://github.com/dotnet/sdk/blob/13b8781fe194a26bc51f1bf766a2f85c04e6298a/src/BuiltInTools/BrowserRefresh/BlazorHotReload.js which is part of the aspnetcore browser reload tools and same bits in Blazor https://github.com/dotnet/aspnetcore/blob/6a93078ee02d3e2aa8ab196ae8911e8bdb803a7b/src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs#L45

File: src/mono/wasm/host/DevServer/DevServerStartup.cs (hunk 5 above)
Comment (maraf) [reply src/mono/wasm/host/DevServer/DevServerStartup.cs:43]: I don't think we have any files that we want to not apply cors on

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets [hunk 6]
```diff
@@ -11,6 +11,19 @@ Copyright (c) .NET Foundation. All rights reserved.
...
 <Project ToolsVersion="14.0">
 
+  <PropertyGroup>
+    <_UseBlazorDevServer>$(RunArguments.Contains('blazor-devserver.dll').ToString().ToLower())</_UseBlazorDevServer>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(_WebAssemblyUserRunParameters)' == '' and '$(_UseBlazorDevServer)' == 'false'">
+    <RunCommand Condition="'$(DOTNET_HOST_PATH)' != '' and Exists($(DOTNET_HOST_PATH))">$(DOTNET_HOST_PATH)</RunCommand>
+    <RunCommand Condition="'$(RunCommand)' == ''">dotnet</RunCommand>
+
+    <WasmAppHostDir>$(MSBuildThisFileDirectory)..\WasmAppHost</WasmAppHostDir>
```
Comment (radical) [inline src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets:21]: Use `$([MSBuild]::NormalizeDirectory`.

File: src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/AppTestBase.cs [hunk 7]
```diff
@@ -57,8 +57,8 @@ protected ToolCommand CreateDotNetCommand() => new DotNetCommand(s_buildEnv, _te
 
     protected async Task<RunResult> RunSdkStyleApp(RunOptions options)
     {
-        string runArgs = $"{s_xharnessRunnerCommand} wasm webserver --app=. --web-server-use-default-files";
-        string workingDirectory = Path.GetFullPath(Path.Combine(FindBlazorBinFrameworkDir(options.Configuration, forPublish: options.ForPublish), ".."));
+        string runArgs = $"run -c {options.Configuration} -p:GenerateRuntimeConfigurationFiles=true";
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/AppTestBase.cs:60]: Add a `FIXME` to remove the property override when the sdk fix flows in, because setting it like this forces the value to be `true`, and the targets files can't change it. thus not testing the same thing as a user would see.

File: src/mono/wasm/host/WebServerStartup.cs [hunk 8]
```diff
@@ -168,41 +168,6 @@ static int GetNextRandomExcept(Range range, params int[] except)
...
-            }
-
-            static string? GetHttpServerAddress(ICollection<string> addresses, bool secure)
-                => addresses?
-                        .Where(a => a.StartsWith(secure ? "https:" : "http:", StringComparison.InvariantCultureIgnoreCase))
-                        .Select(a => new Uri(a))
-                        .Select(uri => uri.ToString())
-                        .FirstOrDefault();
-        });
+        ServerURLsProvider.Hook(app, logger, applicationLifetime, realUrlsAvailableTcs);
```
Comment (radical) [inline src/mono/wasm/host/WebServerStartup.cs:171]: Why the move to an extension method? Is this meant to be used somewhere else too?
If it has to be moved, then the name needs to be descriptive.

File: src/mono/wasm/host/DevServer/DevServerStartup.cs (hunk 5 above)
Comment (radical) [reply src/mono/wasm/host/DevServer/DevServerStartup.cs:43]: What's special about the two files?

File: src/mono/wasm/host/WebServerStartup.cs (hunk 8 above)
Comment (maraf) [reply src/mono/wasm/host/WebServerStartup.cs:171]: It is used from devserver as well. Method renamed

File: src/mono/wasm/host/DevServer/DevServerStartup.cs (hunk 5 above)
Comment (maraf) [reply src/mono/wasm/host/DevServer/DevServerStartup.cs:43]: Blazor has this condition in code that was shared between blazor server and webassembly. And when I added support for COPS to devserver, I duplicated the same condition, but it is probably not needed there.

File: src/mono/wasm/host/DevServer/DebugProxyLauncher.cs (hunk 4 above)
Comment (maraf) [reply src/mono/wasm/host/DevServer/DebugProxyLauncher.cs]: Fixed

File: src/mono/wasm/host/BrowserHost.cs [hunk 9]
```diff
@@ -88,29 +89,104 @@ private async Task RunAsync(ILoggerFactory loggerFactory, CancellationToken toke
...
+        return Environment.CurrentDirectory;
+    }
+
+    private static string? FindFirstFileWithExtension(string directory, string extension)
+    {
+        string[] files = Directory.EnumerateFiles(directory, "*" + extension).ToArray();
+        if (files.Length == 0)
+            return null;
+
+        return files[0];
```
Comment (radical) [inline src/mono/wasm/host/BrowserHost.cs:187]: `=> Directory.EnumerateFiles(..).FirstOrDefault();`

File: src/mono/wasm/host/BrowserHost.cs [hunk 10]
```diff
@@ -88,29 +89,104 @@ private async Task RunAsync(ILoggerFactory loggerFactory, CancellationToken toke
...
+    private static DevServerOptions CreateDevServerOptions(string[] urls, string staticWebAssetsPath, Func<WebSocket, Task>? onConsoleConnected) => new
+    (
+        OnConsoleConnected: onConsoleConnected,
+        StaticWebAssetsPath: staticWebAssetsPath,
+        WebServerUseCors: true,
+        WebServerUseCrossOriginPolicy: true,
+        Urls: urls
+    );
+
+    private static string GetTargetDirectory(string? runtimeConfigPath)
```
Comment (radical) [inline src/mono/wasm/host/BrowserHost.cs:168]: nit: better as a local function in `TryCreateDevServerOptions`, since `target directory`.

File: src/mono/wasm/host/BrowserHost.cs [hunk 11]
```diff
@@ -88,29 +89,104 @@ private async Task RunAsync(ILoggerFactory loggerFactory, CancellationToken toke
...
+    {
+        if (!string.IsNullOrEmpty(runtimeConfigPath))
+        {
+            runtimeConfigPath = Path.GetFullPath(runtimeConfigPath);
+            string? targetPath = Path.GetDirectoryName(runtimeConfigPath);
+            if (!string.IsNullOrEmpty(targetPath))
+                return targetPath;
+        }
+
+        return Environment.CurrentDirectory;
```
Comment (radical) [inline src/mono/wasm/host/BrowserHost.cs:178]: Why this? Rename the method to be clearer about what this method is trying to find.

File: src/mono/wasm/host/BrowserHost.cs [hunk 12]
```diff
@@ -88,29 +89,104 @@ private async Task RunAsync(ILoggerFactory loggerFactory, CancellationToken toke
...
-        );
-
-        (ServerURLs serverURLs, IWebHost host) = await WebServer.StartAsync(options, _logger, token);
+        // If we are using new browser template, use dev server
+        if (TryCreateDevServerOptions(args, urls, onConsoleConnected, out var devServerOptions))
+            return await DevServer.DevServer.StartAsync(devServerOptions, _logger, token);
+
+        // Otherwise for old template, use web server
+        WebServerOptions webServerOptions = CreateWebServerOptions(urls, args.CommonConfig.AppPath, onConsoleConnected);
+        (ServerURLs serverURLs, IWebHost host) = await WebServer.StartAsync(webServerOptions, _logger, token);
```
Comment (radical) [inline src/mono/wasm/host/BrowserHost.cs:107]: It is kinda unclear on why/how we would decide between devserver and the webserver. The code reads as "if creating devserver options" fails for any reason then we fall back to webserver, which would not be correct.

Instead, I would suggest extracting the code from `TryCreateDevServerOptions` to find the static webassets manifest to a separate method, and wrap that in a `IsProjectUsingWasmSdk` (or whatever is relevant). Then the above code can be straight forward to use `DevServer` if the project is using WasmSdk, and WebServer if it is the old style.

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets [hunk 13]
```diff
@@ -11,6 +11,19 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+    <_UseBlazorDevServer>$(RunArguments.Contains('blazor-devserver.dll').ToString().ToLower())</_UseBlazorDevServer>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(_WebAssemblyUserRunParameters)' == '' and '$(_UseBlazorDevServer)' == 'false'">
+    <RunCommand Condition="'$(DOTNET_HOST_PATH)' != '' and Exists($(DOTNET_HOST_PATH))">$(DOTNET_HOST_PATH)</RunCommand>
+    <RunCommand Condition="'$(RunCommand)' == ''">dotnet</RunCommand>
+
+    <WasmAppHostDir>$([MSBuild]::NormalizeDirectory($(MSBuildThisFileDirectory), '..', 'WasmAppHost'))</WasmAppHostDir>
+    <_RuntimeConfigJsonPath>$([MSBuild]::NormalizePath($(OutputPath), '$(AssemblyName).runtimeconfig.json'))</_RuntimeConfigJsonPath>
+    <RunArguments>exec &quot;$([MSBuild]::NormalizePath($(WasmAppHostDir), 'WasmAppHost.dll'))&quot; --runtime-config &quot;$(_RuntimeConfigJsonPath)&quot; $(WasmHostArguments)</RunArguments>
+    <RunWorkingDirectory>$(OutputPath)</RunWorkingDirectory>
```
Comment (radical) [inline src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets:24]: This can be tricky to get correct. How did you test it? @ilonatommy has tested `dotnet run` wasmapphost before too.

File: src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/SatelliteLoadingTests.cs [hunk 14]
```diff
@@ -27,9 +27,8 @@ public SatelliteLoadingTests(ITestOutputHelper output, SharedBuildPerTestClassFi
     public async Task LoadSatelliteAssembly()
     {
         CopyTestAsset("WasmBasicTestApp", "SatelliteLoadingTests");
-        PublishProject("Debug");
 
-        var result = await RunSdkStyleApp(new(Configuration: "Debug", ForPublish: true, TestScenario: "SatelliteAssembliesTest"));
+        var result = await RunSdkStyleApp(new(Configuration: "Debug", TestScenario: "SatelliteAssembliesTest"));
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/SatelliteLoadingTests.cs:31]: Why the change?

File: src/mono/wasm/host/BrowserHost.cs [hunk 15]
```diff
@@ -35,7 +38,6 @@ public BrowserHost(BrowserArguments args, ILogger logger)
                                               CancellationToken token)
     {
         var args = new BrowserArguments(commonArgs);
-        args.Validate();
```
Comment (radical) [inline src/mono/wasm/host/BrowserHost.cs:38]: nit: keep this, and just let that method be empty.

File: src/mono/wasm/host/BrowserArguments.cs [hunk 16]
```diff
@@ -36,9 +36,4 @@ public void ParseJsonProperties(IDictionary<string, JsonElement>? properties)
         if (properties?.TryGetValue("forward-console", out JsonElement forwardConsoleElement) == true)
             ForwardConsoleOutput = forwardConsoleElement.GetBoolean();
     }
-
-    public void Validate()
-    {
-        CommonConfiguration.CheckPathOrInAppPath(CommonConfig.AppPath, HTMLPath, "html-path");
```
Comment (radical) [inline src/mono/wasm/host/BrowserArguments.cs:42]: If this is not null, then would it make sense to still check the path?
Is there a good way to check if the project is using wasm-sdk or not? And if it's not then check this?

File: src/mono/wasm/host/BrowserHost.cs (hunk 11 above)
Comment (maraf) [reply src/mono/wasm/host/BrowserHost.cs:178]: I have dropped the method and used `CommonConfiguration.AppPath`

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets (hunk 13 above)
Comment (maraf) [reply src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets:24]: I played around with simple echo and printing out various variables in various places. Later I added `Debugger.Break()` at the beginning of wasmapphost to be able to debug arguments etc. Is that what your question was about?

File: src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/SatelliteLoadingTests.cs (hunk 14 above)
Comment (maraf) [reply src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/SatelliteLoadingTests.cs:31]: Because instead of publishing the app and using xharness webserver, I moved the tests to `dotnet run`.
I have updated other places in TestAppScenarios

File: src/mono/wasm/host/BrowserArguments.cs (hunk 16 above)
Comment (maraf) [reply src/mono/wasm/host/BrowserArguments.cs:42]: The thing is StaticWebAssets doesn't copy assets to output until you publish the app. Instead they use the generated manifest and serve files from several folders. So the path can be set from WasmSDK project, but the file would not be there (we would have to read the SWA manifest and check correct source folder, but I think it's not worth it).

We can differentiate between old-style and WasmSDK project by passing argument to the WasmAppHost, but my original approach was just to check for SWA manifest. 

Do you think it would be better to have extra arg to clearly switch between project styles? Also note that we will (not sure if in .NET 8) migrate the console template to WasmSDK

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets (hunk 13 above)
Comment (radical) [reply src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets:24]: No, I meant, running the app with `dotnet run`, and with `--no-build`, and running it from the project directory, and outside.

File: src/mono/wasm/host/BrowserArguments.cs (hunk 16 above)
Comment (radical) [reply src/mono/wasm/host/BrowserArguments.cs:42]: > Do you think it would be better to have extra arg to clearly switch between project styles? Also note that we will (not sure if in .NET 8) migrate the console template to WasmSDK

Since we construct the arguments passed to `WasmAppHost`, we can add a parameter for this. This would also allow us to cleanly choose which server to use, whenever we migrate the templates.

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets (hunk 13 above)
Comment (ilonatommy) [reply src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets:24]: Scenarios:
- Create a new WebAssembly Browser App on VS and run, change WasmAppDir and run again.
- Create a new WebAssembly Console App on VS and run, change WasmAppDir and run again.
- Create a new wasmbrowser app using command line and run;
- Create a new wasmconsole app using command line and run;
- Create a new wasmbrowser app using command line and dotnet build and dotnet run --no-build;
- Create a new wasmbrowser app using command line and dotnet build and dotnet run --debug;

File: src/mono/wasm/host/BrowserHost.cs (hunk 12 above)
Comment (maraf) [reply src/mono/wasm/host/BrowserHost.cs:108]: Resolved by specifying additional parameter, as suggested below

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets (hunk 13 above)
Comment (maraf) [reply src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets]: The `WasmAppDir` not relevant for WasmSDK. Tested the rest.

**Review by radical — APPROVED** [review]
Looks great!

File: src/mono/wasm/host/BrowserHost.cs [hunk 17]
```diff
@@ -85,32 +87,104 @@ private async Task RunAsync(ILoggerFactory loggerFactory, CancellationToken toke
...
+    (
+        OnConsoleConnected: onConsoleConnected,
+        StaticWebAssetsPath: staticWebAssetsPath,
+        WebServerUseCors: true,
+        WebServerUseCrossOriginPolicy: true,
+        Urls: urls
+    );
+
+    private static string? FindFirstFileWithExtension(string directory, string extension)
+        => Directory.EnumerateFiles(directory, "*" + extension).First();
```
Comment (radical) [inline src/mono/wasm/host/BrowserHost.cs:186]: It would be useful to throw an exception with details what this tried to look for, and failed. It will be useful for debugging issues.

Comment (maraf) [reply src/mono/wasm/host/BrowserHost.cs:186]: I'll do it in a follow up
